src/core/electrostatics/icc.cpp

```diff
@@ -0,0 +1,283 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/** \file
+ *  Functions to compute the electric field acting on the induced charges,
+ *  excluding forces other than the electrostatic ones. Detailed information
+ *  about the ICC* method is included in the corresponding header file
+ *  \ref icc.hpp.
+ */
+
+#include "config.hpp"
+
+#ifdef ELECTROSTATICS
+
+#include "icc.hpp"
+
+#include "CellStructure.hpp"
+#include "Particle.hpp"
+#include "ParticleRange.hpp"
+#include "actor/visitors.hpp"
+#include "cells.hpp"
+#include "communication.hpp"
+#include "electrostatics/coulomb.hpp"
+#include "electrostatics/coulomb_inline.hpp"
+#include "errorhandling.hpp"
+#include "event.hpp"
+#include "integrate.hpp"
+
+#include <utils/constants.hpp>
+
+#include <boost/mpi/collectives/all_reduce.hpp>
+#include <boost/mpi/operations.hpp>
+
+#include <algorithm>
+#include <cassert>
+#include <cmath>
+#include <stdexcept>
+#include <vector>
+
+/** Calculate the electrostatic forces between source charges (= real charges)
+ *  and wall charges. For each electrostatic method, the proper functions
+ *  for short- and long-range parts are called. Long-range parts are calculated
+ *  directly, short-range parts need helper functions according to the particle
+ *  data organisation. This is a modified version of \ref force_calc.
+ */
+static void force_calc_icc(CellStructure &cell_structure,
+                           ParticleRange const &particles,
+                           ParticleRange const &ghost_particles) {
+  // reset forces
+  for (auto &p : particles) {
+    p.force() = {};
+  }
+  for (auto &p : ghost_particles) {
+    p.force() = {};
+  }
+
+  auto const kernel = Coulomb::central_force_kernel();
+
+  // calc ICC forces
+  cell_structure.non_bonded_loop(
+      [&kernel](Particle &p1, Particle &p2, Distance const &d) {
+        auto const q1q2 = p1.q() * p2.q();
+        if (q1q2 != 0.) {
+          auto force = kernel(q1q2, d.vec21, std::sqrt(d.dist2));
+          p1.force() += force;
+          p2.force() -= force;
+#ifdef P3M
+          if (auto actor = get_actor_by_type<ElectrostaticLayerCorrection>(
+                  electrostatics_actor)) {
+            actor->add_pair_force_corrections(p1, p2, q1q2);
+          }
+#endif // P3M
+        }
+      });
+
+  Coulomb::calc_long_range_force(particles);
+}
+
+void ICCStar::iteration(CellStructure &cell_structure,
+                        ParticleRange const &particles,
+                        ParticleRange const &ghost_particles) {
+  if (icc_cfg.n_icc == 0)
+    return;
+
+  try {
+    sanity_check();
+  } catch (std::runtime_error const &err) {
+    runtimeErrorMsg() << err.what();
+    return;
+  }
+
+  auto prefactor = 0.;
+  if (electrostatics_actor) {
+    prefactor =
+        boost::apply_visitor(GetCoulombPrefactor{}, *electrostatics_actor);
+  }
+  assert(prefactor != 0.);
```

Comment:
Just out of interest, is there a way to reach that point with no `electrostatics_actor` activated? In my ideal world ICC*-iterations should never be executed without an electrostatics solver. That would make the `if (electrostatics_actor)` to always evalute to `true`, right?